R/dataframe-frame.R

```diff
@@ -1698,3 +1698,60 @@ dataframe__rolling <- function(
     )
   })
 }
+
+#' Transpose a DataFrame over the diagonal
+#'
+#' @inheritParams rlang::args_dots_empty
+#' @param include_header If set, the column names will be added as first column.
+#' @param header_name If `include_header` is set, this determines the name of
+#' the column that will be inserted.
+#' @param column_names Optional string naming an existing column, or a function
+#' that takes an integer vector representing the position of value (non-header)
+#' columns and returns a character vector of same length. Column position is
+#' 0-indexed.
+#'
+#' @inherit as_polars_df return
+#'
+#' @details
+#' This is a very expensive operation. Perhaps you can do it differently.
+#'
+#' @examples
+#' df <- pl$DataFrame(a = c(1, 2, 3), b = c(4, 5, 6))
+#' df$transpose(include_header = TRUE)
+#'
+#' # Replace the auto-generated column names with a list
+#' df$transpose(include_header = FALSE, column_names = c("x", "y", "z"))
+#'
+#' # Include the header as a separate column
+#' df$transpose(
+#'   include_header = TRUE, header_name = "foo", column_names = c("x", "y", "z")
+#' )
+#'
+#' # Use a function to produce the new column names
+#' name_generator <- function(x) {
+#'   paste0("my_column_", x)
+#' }
+#' df$transpose(include_header = FALSE, column_names = name_generator)
+#'
+#' # Use an existing column as the new column names
+#' df <- pl$DataFrame(id = c("i", "j", "k"), a = c(1, 2, 3), b = c(4, 5, 6))
+#' df$transpose(column_names = "id")
+#' df$transpose(include_header = TRUE, header_name = "new_id", column_names = "id")
+dataframe__transpose <- function(
+    ...,
+    include_header = FALSE,
+    header_name = "column",
+    column_names = NULL) {
+  wrap({
+    check_dots_empty0(...)
+    keep_names_as <- if (isTRUE(include_header)) {
+      header_name
+    } else {
+      NULL
+    }
+    if (is.function(column_names)) {
+      column_names <- column_names(seq_len(nrow(self)) - 1)
+    }
+    self$`_df`$transpose(keep_names_as = keep_names_as, column_names = column_names %||% character())
```

Comment:
If you give the `header_name` something other than a string or NULL, the name displayed in the error here would be `keep_name_as`.
So the argument name of the Rust side should be `header_name`

Reply:
Or should we just add `check_character(header_name, allow_null = TRUE)`?

Reply:
Actually that wouldn't take into account length errors. I'll just rename the arg in rut

Reply:
Yes, using `check_string` here is fine.

Reply:
Ah I forgot we had `check_string()` on top of `check_character()`